File: backend/src/main/java/com/ultron/intelligence/psychology/PriorityScorer.java
package com.ultron.intelligence.psychology;

import java.util.Locale;
import org.springframework.stereotype.Component;

/**
 * Priority scoring (L7 — The Mind). Produces a 0..1 importance score for an item so the Planner
 * and proactive layers can order work. Offline + deterministic in Phase 1 (keyword + signal
 * heuristics); swappable for the XGBoost model in the Python layer later.
 */
@Component
public class PriorityScorer {

    private static final String[] URGENT = {"urgent", "asap", "immediately", "now", "critical", "emergency", "today", "deadline"};
    private static final String[] HIGH = {"important", "must", "blocker", "blocked", "overdue", "high priority"};
    private static final String[] LOW = {"someday", "maybe", "whenever", "low priority", "nice to have", "fyi"};

    /** Score in [0,1]. Higher = more urgent/important. */
    public double score(String text) {
        if (text == null || text.isBlank()) {
            return 0.0;
        }
        String t = text.toLowerCase(Locale.ROOT);
        double score = 0.3; // neutral baseline

        for (String w : URGENT) {
            if (t.contains(w)) {
                score += 0.25;
            }
        }
        for (String w : HIGH) {
            if (t.contains(w)) {
                score += 0.15;
            }
        }
        for (String w : LOW) {
            if (t.contains(w)) {
                score -= 0.2;
            }
        }
        if (t.contains("!")) {
            score += 0.05;
        }
        return clamp(score);
    }

    private static double clamp(double v) {
        return Math.max(0.0, Math.min(1.0, v));
    }
}

File: ai-layer/psychology/priority_scorer.py
"""Priority scoring (L7). Stdlib heuristic mirroring the Java PriorityScorer.

Upgradeable to an XGBoost model behind the same ``score`` surface.
"""
from __future__ import annotations

_URGENT = ("urgent", "asap", "immediately", "now", "critical", "emergency", "today", "deadline")
_HIGH = ("important", "must", "blocker", "blocked", "overdue", "high priority")
_LOW = ("someday", "maybe", "whenever", "low priority", "nice to have", "fyi")


def score(text: str) -> float:
    if not text or not text.strip():
        return 0.0
    t = text.lower()
    s = 0.3
    s += 0.25 * sum(1 for w in _URGENT if w in t)
    s += 0.15 * sum(1 for w in _HIGH if w in t)
    s -= 0.20 * sum(1 for w in _LOW if w in t)
    if "!" in t:
        s += 0.05
    return max(0.0, min(1.0, s))
